File: backend/models/whiteboards.py
"""
Whiteboard Collaboration Models
Pydantic models for whiteboard management and collaboration
"""

from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime


class WhiteboardCreate(BaseModel):
    """Model for creating a new whiteboard"""
    title: str = Field(..., min_length=1, max_length=200)
    content: Optional[str] = Field(default='{"elements":[],"appState":{}}', description="Serialized Excalidraw scene")
    collaborators: Optional[List[str]] = Field(default_factory=list, description="Array of user emails")
    is_public: bool = Field(default=False, alias="isPublic")

    class Config:
        populate_by_name = True


class WhiteboardUpdate(BaseModel):
    """Model for updating an existing whiteboard"""
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    content: Optional[str] = Field(None, description="Serialized Excalidraw scene")
    collaborators: Optional[List[str]] = Field(None, description="Array of user emails")
    is_public: Optional[bool] = Field(None, alias="isPublic")
    updated_at: Optional[str] = Field(None, alias="updatedAt")

    class Config:
        populate_by_name = True


class WhiteboardResponse(BaseModel):
    """Model for whiteboard response"""
    id: str = Field(alias="$id")
    title: str
    content: str
    created_by: str = Field(alias="createdBy")
    created_at: str = Field(alias="createdAt")
    updated_at: str = Field(alias="updatedAt")
    collaborators: List[str]
    is_public: bool = Field(alias="isPublic")

    class Config:
        populate_by_name = True


class WhiteboardListResponse(BaseModel):
    """Model for whiteboard list response"""
    documents: List[WhiteboardResponse]
    total: int


class WhiteboardListItem(BaseModel):
    """Simplified whiteboard info for list views"""
    id: str = Field(alias="$id")
    title: str
    created_by: str = Field(alias="createdBy")
    updated_at: str = Field(alias="updatedAt")
    collaborators: List[str]
    is_owner: bool = Field(alias="isOwner")

    class Config:
        populate_by_name = True


# Whiteboard Invites
class WhiteboardInviteCreate(BaseModel):
    """Model for creating a whiteboard invite"""
    whiteboard_id: str = Field(alias="whiteboardId")
    email: str
    invited_by: str = Field(alias="invitedBy")

    class Config:
        populate_by_name = True


class WhiteboardInviteResponse(BaseModel):
    """Model for whiteboard invite response"""
    id: str = Field(alias="$id")
    whiteboard_id: str = Field(alias="whiteboardId")
    email: str
    invited_by: str = Field(alias="invitedBy")
    status: str  # 'pending' | 'accepted' | 'rejected'
    token: str
    created_at: str = Field(alias="createdAt")
    expires_at: str = Field(alias="expiresAt")

    class Config:
        populate_by_name = True


class WhiteboardInviteListResponse(BaseModel):
    """Model for whiteboard invite list response"""
    documents: List[WhiteboardInviteResponse]
    total: int


# Whiteboard Presence
class WhiteboardPresenceCreate(BaseModel):
    """Model for creating whiteboard presence tracking"""
    whiteboard_id: str = Field(alias="whiteboardId")
    user_email: str = Field(alias="userEmail")
    color: str

    class Config:
        populate_by_name = True


class WhiteboardPresenceResponse(BaseModel):
    """Model for whiteboard presence response"""
    id: str = Field(alias="$id")
    whiteboard_id: str = Field(alias="whiteboardId")
    user_email: str = Field(alias="userEmail")
    color: str
    last_seen: str = Field(alias="lastSeen")
    is_active: bool = Field(alias="isActive")

    class Config:
        populate_by_name = True

File: frontend/types/whiteboards.ts
/**
 * Type definitions for the Whiteboard Collaboration System
 */

/**
 * Main whiteboard model stored in Appwrite
 */
export interface WhiteboardModel {
  $id: string;
  title: string;
  content: string; // Serialized Excalidraw scene data
  createdBy: string; // User email
  createdAt: string; // ISO datetime
  updatedAt: string; // ISO datetime
  collaborators: string[]; // Array of user emails
  isPublic: boolean;
}

/**
 * Whiteboard invite model
 */
export interface WhiteboardInvite {
  $id: string;
  whiteboardId: string;
  email: string;
  invitedBy: string;
  status: 'pending' | 'accepted' | 'rejected';
  token: string;
  createdAt: string;
  expiresAt: string;
}

/**
 * Real-time presence tracking for whiteboard users
 */
export interface WhiteboardPresence {
  $id: string;
  whiteboardId: string;
  userEmail: string;
  color: string; // Hex color for cursor/selection
  lastSeen: string;
  isActive: boolean;
}

/**
 * Simplified whiteboard info for list views
 */
export interface WhiteboardListItem {
  $id: string;
  title: string;
  createdBy: string;
  updatedAt: string;
  collaborators: string[];
  isOwner: boolean;
  hasUnreadInvites?: boolean;
}

/**
 * Parameters for creating a new whiteboard
 */
export interface CreateWhiteboardParams {
  title: string;
  initialContent?: string;
  collaborators?: string[];
}

/**
 * Collaborator information
 */
export interface CollaboratorInfo {
  email: string;
  color: string;
  isActive: boolean;
}

/**
 * Parameters for accepting an invite
 */
export interface AcceptWhiteboardInviteParams {
  token: string;
  userEmail: string;
}

/**
 * Parameters for rejecting an invite
 */
export interface RejectWhiteboardInviteParams {
  token: string;
  userEmail: string;
}
